### PR TITLE
Gives the comms agent a megaphone

### PR DESCRIPTION
**Pull request by CydiaLamiales**
## About The Pull Request
Gives the comms agent a megaphone.
Makes the megaphones have a defined cooldown.

## Why It's Good For The Game
Allows for communication agents to impersonate command staff.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

Comms agent starting with a megaphone

image

</details>

## Changelog
:cl:
add: Gave the Lavaland Comms Agent a megaphone with 5 "charges" and a two minute recharge time for each use.

### CODE REVIEW

**Comment by Fronsis** [comment]
I feel like i agree to what Bacon said on the other PR, the most annoying and spammer character below the clown is the syndie comms agent, while it might be good to impersonate from time to time a command officer i feel like the other 90% of the time it's going to be loud mode saying nonsense and compared to the VIP/Clown (who you can arrest and take it off them) you'll need a lot of preparation to assault that base, even if you specifically target their t-comms array with a bomb or something

Considering that they now have a higher spawn rate it's quite annoying to have an unfilterable spam of someone being ''rude and obnoxious'' and the reality is.. only a few of them do an actual good job of trying to cause chaos before resorting to bickering the general or sec comms

**Comment by CydiaLamiales** [comment]
That's why get they get a version with a 30 second coolsown between messages, so they can't just spam it.

**Comment by Rukofamicom** [comment]
Once every 30 seconds is still quite spammable

**Review by EvilDragonfiend — CHANGES_REQUESTED** [review]
* Megaphone should tell how much time you should wait to reuse that.
* Increase the cooldown time to between 90-120s, or even it should be longer than this.
* If you want to make it spammable in short period, charge version would work. (like it's charged for ever 2 minutes, and maximum charge is 5.) This would make the thing you want (impersonate a head) while not allowing them spamming this too much.
other method to recharge is that it can be fully recharged for every 10 minutes.

I wonder if you have ever seen my VIP gimmick. People get ear damage extremely when I say a lot of stuff in short time in loudmode.. I like comms agent being annoying, but they shouldn't disrupt the radio every 30 seconds. it's still too short.

**Comment by Rukofamicom** [comment]
5 charges, two minutes per charge seems like a nice balance between good impersonation and spam prevention.

I don't believe the cooldowns or charges need to be readily visible to the player though. I even think it would be better if they had a permanent burnout function like security masks if overused, with a vague warning message when close. (This is not a requested change, just some thoughts)

**Review by Rukofamicom — APPROVED** [review]

File: code/game/objects/items/devices/megaphone.dm [hunk 1]
```diff
@@ -65,3 +66,54 @@
 	icon_state = "megaphone-clown"
 	item_state = "megaphone-clown"
 	voicespan = list(SPAN_CLOWN)
+
+/obj/item/megaphone/nospam
+	cooldown = 2 MINUTES // So it can be varedited if needed
+	var/list/charges_list = list(0,0,0,0,0)
+	var/available_charges = 5
```
Comment (EvilDragonfiend) [inline code/game/objects/items/devices/megaphone.dm]:
```suggestion
	var/list/charges_list = list()
	var/maximum_charge = 5
```
make it like this.

File: code/game/objects/items/devices/megaphone.dm [hunk 2]
```diff
@@ -65,3 +66,54 @@
...
+/obj/item/megaphone/nospam/process(delta_time)
+	if(available_charges >= 5)
+		return
+	var/I = 1
+	var/charges = 0
+	while(I < 6)
+		if(charges_list[I] < world.time)
+			charges ++
+		I++
+	available_charges = charges
```
Comment (EvilDragonfiend) [inline code/game/objects/items/devices/megaphone.dm]:
```suggestion
	var/current_index = length(charges_list)
	while(current_index > 0)
		if(charges_list[current_index] < world.time)
			charges_list.Remove(current_index)
		current_index--
```

**Review by EvilDragonfiend — CHANGES_REQUESTED** [review]
I believe these two suggestions are enough to tell what I am thinking. Ping me if you're not sure.

File: code/game/objects/items/devices/megaphone.dm [hunk 3]
```diff
@@ -65,3 +66,54 @@
...
+	if (user.get_active_held_item() != src)
+		return
+	if(available_charges > 0)
+		var/I = 1
+		while(I < 6)
+			if(charges_list[I] < world.time)
+				charges_list[I] = world.time + cooldown
+				available_charges --
+				break
+			I++
```
Comment (EvilDragonfiend) [inline code/game/objects/items/devices/megaphone.dm]:
```suggestion
	if(length(charges_list) < maximum_charge)
		charges_list.Add(world.time + cooldown)
```

**Review by EvilDragonfiend — CHANGES_REQUESTED** [review]

File: code/game/objects/items/devices/megaphone.dm [hunk 4]
```diff
@@ -65,3 +66,45 @@
...
+
+/obj/item/megaphone/nospam/Initialize(mapload)
+	. = ..()
+	START_PROCESSING(SSobj, src)
+
+/obj/item/megaphone/nospam/process(delta_time)
+	var/current_index = length(charges_list)
+	while(current_index > 0)
+		if(charges_list[current_index] < world.time)
+			charges_list.Remove(current_index)
```
Comment (EvilDragonfiend) [inline code/game/objects/items/devices/megaphone.dm]:
```suggestion
			charges_list.Cut(current_index, current_index+1)
```

**Review by EvilDragonfiend — CHANGES_REQUESTED** [review]

**Review by EvilDragonfiend — APPROVED** [review]
Maybe `addtimer()` would have been better but I am not sure which one is expensive. I will just merge it anyway.

**Pull request merged**